lib/lints/avoid_unnecessary_type_assertions/avoid_unnecessary_type_assertions_rule.dart

```diff
@@ -0,0 +1,165 @@
+
+class AvoidUnnecessaryTypeAssertions extends SolidLintRule {
+  /// The [LintCode] of this lint rule that represents
+  /// the error whether we use bad formatted double literals.
+  static const lintName = 'avoid-unnecessary-type-assertions';
+
+  static const _unnecessaryIsCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'is' operator.",
+  );
+
+  static const _unnecessaryWhereTypeCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'whereType' method.",
+  );
+
+  AvoidUnnecessaryTypeAssertions._(super.config);
+
+  /// Creates a new instance of [AvoidUnnecessaryTypeAssertions]
+  /// based on the lint configuration.
+  factory AvoidUnnecessaryTypeAssertions.createRule(CustomLintConfigs configs) {
+    final rule = RuleConfig(
+      configs: configs,
+      name: lintName,
+      problemMessage: (_) => "Unnecessary usage of typecast operators.",
+    );
+
+    return AvoidUnnecessaryTypeAssertions._(rule);
+  }
+
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addIsExpression((node) {
+      if (_isUnnecessaryIsExpression(node)) {
+        reporter.reportErrorForNode(_unnecessaryIsCode, node);
+      }
+    });
+
+    context.registry.addMethodInvocation((node) {
+      if (_isUnnecessaryWhereType(node)) {
+        reporter.reportErrorForNode(_unnecessaryWhereTypeCode, node);
+      }
+    });
+  }
+
+  @override
+  List<Fix> getFixes() => [_UnnecessaryTypeAssertionsFix()];
+
+  bool _isUnnecessaryIsExpression(IsExpression node) {
+    final objectType = node.expression.staticType;
+    final castedType = node.type.type;
+
+    if (node.notOperator != null &&
+        objectType != null &&
+        objectType is! TypeParameterType &&
+        objectType is! DynamicType &&
+        !objectType.isDartCoreObject &&
+        _isUnnecessaryTypeCheck(objectType, castedType, isReversed: true)) {
+      return true;
+    } else {
+      return _isUnnecessaryTypeCheck(objectType, castedType);
+    }
+  }
+
+  bool _isUnnecessaryWhereType(MethodInvocation node) {
+    const whereTypeMethodName = 'whereType';
+
+    final targetType = node.target?.staticType;
+
+    if (node.methodName.name == whereTypeMethodName &&
+        isIterable(node.realTarget?.staticType) &&
+        targetType is ParameterizedType) {
+      final targetHasExplicitTypeArgs = targetType.typeArguments.isNotEmpty;
+      final arguments = node.typeArguments?.arguments;
+      final invocationHasExplicitTypeArgs = arguments?.isNotEmpty ?? false;
+
+      return targetHasExplicitTypeArgs &&
+          invocationHasExplicitTypeArgs &&
+          _isUnnecessaryTypeCheck(
+            targetType.typeArguments.first,
+            arguments?.first.type,
+          );
+    }
+    return false;
+  }
+
+  bool _isUnnecessaryTypeCheck(
+    DartType? objectType,
+    DartType? castedType, {
+    bool isReversed = false,
```

Comment:
I find it hard to understand what this flag does in the method.
Need better naming, code comments or functional decomposition.